sql/distsql: TestHashJoinerAgainstProcessor failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/39e43fac9834b338e381641f0f573c060320d678

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestHashJoinerAgainstProcessor PKG=github.com/cockroachdb/cockroach/pkg/sql/distsql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1563910&tab=buildLog

```
=== RUN   TestHashJoinerAgainstProcessor
--- join type = FULL_OUTER onExpr = "" filter = "" seed = 2742747369511917154 run = 2 ---
--- lEqCols = [2] rEqCols = [2] ---
--- inputTypes = [{{BoolFamily 0 0 [] 0x65613d0 0 <nil> [] [] 16 <nil>}} {{DecimalFamily 0 0 [] 0x65613d0 0 <nil> [] [] 1700 <nil>}} {{BytesFamily 0 0 [] 0x65613d0 0 <nil> [] [] 17 <nil>}}] ---
--- FAIL: TestHashJoinerAgainstProcessor (0.54s)
    columnar_operators_test.go:267: unexpected meta &{Ranges:[] Err:assertion failure
          - error with attached stack trace:
            github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror.CatchVectorizedRuntimeError.func1
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:77
            runtime.gopanic
            	/usr/local/go/src/runtime/panic.go:522
            github.com/cockroachdb/cockroach/pkg/col/coldata.(*Bytes).UpdateOffsetsToBeNonDecreasing
            	/go/src/github.com/cockroachdb/cockroach/pkg/col/coldata/bytes.go:75
            github.com/cockroachdb/cockroach/pkg/col/coldata.(*MemBatch).SetLength
            	/go/src/github.com/cockroachdb/cockroach/pkg/col/coldata/batch.go:140
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*hashJoinEqOp).emitUnmatched
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:330
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*hashJoinEqOp).Next
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:267
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).next
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:149
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).nextAdapter
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:140
            github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror.CatchVectorizedRuntimeError
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:91
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:177
            github.com/cockroachdb/cockroach/pkg/sql/distsql.verifyColOperator
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/columnar_utils_test.go:119
            github.com/cockroachdb/cockroach/pkg/sql/distsql.TestHashJoinerAgainstProcessor
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/columnar_operators_test.go:256
            testing.tRunner
            	/usr/local/go/src/testing/testing.go:865
            runtime.goexit
            	/usr/local/go/src/runtime/asm_amd64.s:1337
          - error with embedded safe details: unexpected error from the vectorized runtime: %+v
            -- arg 1: <string>
          - unexpected error from the vectorized runtime: unexpectedly found a decreasing non-zero offset: previous max=32, found=4 TraceData:[] TxnCoordMeta:<nil> RowNum:<nil> SamplerProgress:<nil> BulkProcessorProgress:<nil> Metrics:<nil>} from columnar operator

```

SHA: https://github.com/cockroachdb/cockroach/commits/39e43fac9834b338e381641f0f573c060320d678

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestHashJoinerAgainstProcessor PKG=github.com/cockroachdb/cockroach/pkg/sql/distsql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1563912&tab=buildLog

```
=== RUN   TestHashJoinerAgainstProcessor
--- join type = FULL_OUTER onExpr = "" filter = "@2 > e'\\'gdr\\x1f\\x17\\x1a'" seed = 8834711684691466341 run = 1 ---
--- lEqCols = [0] rEqCols = [0] ---
--- inputTypes = [{{StringFamily 0 0 [] 0x84f6a00 0 <nil> [] [] 19 <nil>}} {{StringFamily 0 0 [] 0x84f6a00 0 <nil> [] [] 1042 <nil>}} {{OidFamily 0 0 [] 0x84f6a00 0 <nil> [] [] 2205 <nil>}}] ---
--- FAIL: TestHashJoinerAgainstProcessor (0.34s)
    columnar_operators_test.go:267: unexpected meta &{Ranges:[] Err:assertion failure
          - error with attached stack trace:
            github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror.CatchVectorizedRuntimeError.func1
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:77
            runtime.gopanic
            	/usr/local/go/src/runtime/panic.go:522
            github.com/cockroachdb/cockroach/pkg/col/coldata.(*Bytes).UpdateOffsetsToBeNonDecreasing
            	/go/src/github.com/cockroachdb/cockroach/pkg/col/coldata/bytes.go:75
            github.com/cockroachdb/cockroach/pkg/col/coldata.(*MemBatch).SetLength
            	/go/src/github.com/cockroachdb/cockroach/pkg/col/coldata/batch.go:140
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*hashJoinEqOp).emitUnmatched
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:330
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*hashJoinEqOp).Next
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:267
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*selGTBytesBytesConstOp).Next
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/selection_ops.eg.go:2698
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).next
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:149
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).nextAdapter
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:140
            github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror.CatchVectorizedRuntimeError
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:91
            github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:177
            github.com/cockroachdb/cockroach/pkg/sql/distsql.verifyColOperator
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/columnar_utils_test.go:119
            github.com/cockroachdb/cockroach/pkg/sql/distsql.TestHashJoinerAgainstProcessor
            	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/columnar_operators_test.go:256
            testing.tRunner
            	/usr/local/go/src/testing/testing.go:865
            runtime.goexit
            	/usr/local/go/src/runtime/asm_amd64.s:1337
          - error with embedded safe details: unexpected error from the vectorized runtime: %+v
            -- arg 1: <string>
          - unexpected error from the vectorized runtime: unexpectedly found a decreasing non-zero offset: previous max=37, found=34 TraceData:[] TxnCoordMeta:<nil> RowNum:<nil> SamplerProgress:<nil> BulkProcessorProgress:<nil> Metrics:<nil>} from columnar operator

```

--- a/pkg/col/coldata/bytes.go
+++ b/pkg/col/coldata/bytes.go
@@ -66,16 +66,10 @@ func (b *Bytes) AssertOffsetsAreNonDecreasing(n uint64) {
 func (b *Bytes) UpdateOffsetsToBeNonDecreasing(n uint64) {
 	prev := b.offsets[0]
 	for j := uint64(1); j <= n; j++ {
-		if b.offsets[j] == 0 {
-			b.offsets[j] = prev
-		} else if b.offsets[j] < prev {
-			// Having a zero offset is expected (when an actual NULL value is there)
-			// whereas a non-zero offset that is smaller than the largest offset seen
-			// so far is unexpected.
-			panic(fmt.Sprintf("unexpectedly found a decreasing non-zero offset:"+
-				" previous max=%d, found=%d", prev, b.offsets[j]))
-		} else {
+		if b.offsets[j] > prev {
 			prev = b.offsets[j]
+		} else {
+			b.offsets[j] = prev
 		}
 	}
 }

